sdk/go/common/resource/plugin/provider-implementers-guide.md

```diff
@@ -0,0 +1,609 @@
+ID of the resource to import; that is, _any importable resource must be identifiable using
+its ID alone_. If the [`Read`](#read) succeeds, the engine calls the provider's [`Check`](#check) method with
+the inputs returned by [`Read`](#read) and the inputs supplied by the user. If any of the inputs
+are invalid, the import fails. Finally, the engine calls the provider's [`Diff`](#diff) method with
+the inputs returned by [`Check`](#check) and the state returned by [`Read`](#read). If the call to [`Diff`](#diff)
+indicates that there is no difference between the desired state described by the inputs
+and the actual state, the import succeeds. Otherwise, the import fails.
+
+If the resource is being managed, the engine first looks up the last registered inputs and
+last refreshed state for the resource's URN. The engine then calls the resource's
+provider's [`Check`](#check) method with the last registered inputs (if any) and the inputs supplied
+by the user. If any of the inputs are invalid, the registration fails. Otherwise, the
+engine decides which operations to perform on the resource based on the difference between
+the desired state described by its inputs and its actual state. If the resource does not
+exist (i.e. there is no last refereshed state for its URN), the engine calls the
+provider's [`Create`](#create) method, which returns the ID and state of the created resource. If the
+resource does exist, the action taken depends on the differences (if any) between the
+desired and actual state of the resource.
+
+If the resource does exist, the engine calls the provider's [`Diff`](#diff) method with the
+inputs returned from [`Check`](#check), the resource's ID, and the resource's last refreshed state.
+If the result of the call indicates that there is no difference between the desired and
+actual state, no operation is necessary. Otherwise, the resource is either updated (if
+[`Diff`](#diff) does not indicate that the resource must be replaced) or replaced (if [`Diff`](#diff) does
+indicate that the resource must be replaced).
+
+To update a resource, the engine calls the provider's [`Update`](#update) method with the inputs
+returned from [`Check`](#check), the resource's ID, and its last refreshed state. [`Update`](#update) returns
+the new state of the resource. The resource's ID may not be changed by a call to [`Update`](#update).
+
+To replace a resource, the engine first calls [`Check`](#check) with an empty set of prior inputs
+and the inputs supplied with the resource's registration. If [`Check`](#check) fails, the resource
+is not replaced. Otherwise, the inputs returned by this call to [`Check`](#check) will be used to
+create the replacement resource. Next, the engine inspects the resource options supplied
+with the resource's registration and result of the call to [`Diff`](#diff) to determine whether
+the replacement can be created before the original resource is deleted. This order of
```

Comment:
Perhaps explicitly note - "if either the `deleteBeforeReplace` resource option or the `deleteBeforeReplace` property of the `Diff` output is true, the resource will be deleted before it's replacement is created, else its replacement will be created before it is deleted."